runners/google-cloud-dataflow-java/src/main/java/org/apache/beam/runners/dataflow/DataflowPipelineJob.java

```diff
@@ -302,11 +302,18 @@ public State cancel() throws IOException {
           .update(projectId, jobId, content)
           .execute();
     } catch (IOException e) {
-      String errorMsg = String.format(
-          "Failed to cancel the job, please go to the Developers Console to cancel it manually: %s",
-          MonitoringUtil.getJobMonitoringPageURL(getProjectId(), getJobId()));
-      LOG.warn(errorMsg);
-      throw new IOException(errorMsg, e);
+      State state = getState();
+      if (state.isTerminal()) {
+        LOG.warn("Job is already terminated. State is {}", state);
+        return state;
+      } else {
+        String errorMsg = String.format(
+            "Failed to cancel the job, "
+                + "please go to the Developers Console to cancel it manually: %s",
+            MonitoringUtil.getJobMonitoringPageURL(getProjectId(), getJobId()));
+        LOG.warn(errorMsg);
+        throw new IOException(errorMsg, e);
+      }
     }
     return State.CANCELLED;
```

Comment:
Move this line to be after the execute(), this will prevent coding bugs where default fallthrough returns cancelled status